### ISSUE TITLE
Cannot connect to Private Gitlab Repo

### ISSUE THREAD

**Issue by tomjohnburton** [argoproj/argo-cd#2890]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

I have successfully connected to my private repo hosted on gitlab using the UI. The UI confirms successful connection. 
![image](https://user-images.githubusercontent.com/39559415/71018254-7a0ef400-20f8-11ea-9f2c-10f5e7a676a9.png)

When trying to create an app I receive the following error
```
Unable to create application: application spec is invalid: 
InvalidSpecError: Unable to get app details: rpc error: code = Internal desc = Failed to fetch git 
repo: `git fetch origin --tags --force` failed exit status 128: 
Failed to add the ECDSA host key for IP address 'XX.XXX.XXX.XX' to the list of known hosts (/app/config/ssh/ssh_known_hosts). 
Load key "/dev/shm/568617259": invalid format git@gitlab.com: Permission denied (publickey). fatal: Could not read from remote repository. 
Please make sure you have the correct access rights and the repository exists.
```

**Expected behavior**

To successfully clone repo

**Version**

```shell
argocd: v1.3.6+89be1c9
  BuildDate: 2019-12-10T22:48:19Z
  GitCommit: 89be1c9ce6db0f727c81277c1cfdfb1e385bf248
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.3.6+89be1c9
  BuildDate: 2019-12-10T22:47:48Z
  GitCommit: 89be1c9ce6db0f727c81277c1cfdfb1e385bf248
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: v2.15.2
  Kubectl Version: v1.14.0
```

**Comment by jengo** [comment]
I am experiencing the exact same issue with Github

**Comment by rickdennis** [comment]
+1 same

**Comment by RobertKozak** [comment]
I'm also experiencing with GitHub:

`argocd: v1.3.6+89be1c9
  BuildDate: 2019-12-10T22:48:19Z
  GitCommit: 89be1c9ce6db0f727c81277c1cfdfb1e385bf248
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.3.6+89be1c9
  BuildDate: 2019-12-10T22:47:48Z
  GitCommit: 89be1c9ce6db0f727c81277c1cfdfb1e385bf248
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: v2.15.2
  Kubectl Version: v1.14.0`

**Comment by philippeboyd** [comment]
I just had the same problem today. Turns out, Argo has a glitch when entering the private key via the UI...

I added my repository via the command line and it works now; I can then create an application successfully.

```bash 
argocd repo add git@github.com:you/your-repo.git --ssh-private-key $HOME/.ssh/private_key
```

https://argoproj.github.io/argo-cd/user-guide/private-repositories/#ssh-private-key-credential

**Comment by scbizu** [comment]
Any updates about this issue ?

**Comment by jannfis** [comment]
For the time being, can you please try to add the repository using a full qualified SSH URL, i.e. `ssh://git@github.com/you/your-repo.git` instead of `git@github.com:you/your-repo.git` and see if that functions as a workaround?

**Comment by scbizu** [comment]
> For the time being, can you please try to add the repository using a full qualified SSH URL, i.e. `ssh://git@github.com/you/your-repo.git` instead of `git@github.com:you/your-repo.git` and see if that functions as a workaround?

Still can not . 

![image](https://user-images.githubusercontent.com/8732481/72770136-ecaf4a00-3c37-11ea-99a8-39a85919793c.png)

**Comment by scbizu** [comment]
BTW, why you try to `git fetch` the origin repo even if I already connected the repo origin to the origin git server rather than read the source file from the local temp git directory ?

**Comment by scbizu** [comment]
And I think that I catch the bug :

As https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L1151  uses the `getCAPath` to fetch the SSH Cred file path , but the `getCAPath` only checks if it is a HTTPS Cred and no options for SSH .

As for the fix , I think it's better to add the other `if` case for the SSH , like this:

```go
func getCAPath(repoURL string) string {
	if git.IsHTTPSURL(repoURL) {
		if parsedURL, err := url.Parse(repoURL); err == nil {
			if caPath, err := cert.GetCertBundlePathForRepository(parsedURL.Host); err == nil {
				return caPath
			} else {
				log.Warnf("Could not get cert bundle path for host '%s'", parsedURL.Host)
			}
		} else {
			// We don't fail if we cannot parse the URL, but log a warning in that
			// case. And we execute the command in a verbatim way.
			log.Warnf("Could not parse repo URL '%s'", repoURL)
		}
	}

     if git.IsSSHURL(repoURL){
           // TODO: fetch the SSH cred file path
     }

	return ""
}
```    

I can send my PR if you guys do not have enough time to solve this bug.

**Comment by jannfis** [comment]
> And I think that I catch the bug :
> 
> As https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L1151 uses the `getCAPath` to fetch the SSH Cred file path , but the `getCAPath` only checks if it is a HTTPS Cred and no options for SSH .

Hm, that function only resolves to the file the TLS CA certificate for HTTPS connected repositories should be loaded from. The SSH private key is actually stored in a Kubernetes secret (and written out to a temporary file during connection phase).

I just tried to reproduce the following:

- Connect new repository via SSH using the UI, using `git@gitlab.com:user/repo` as the repository URL and pasted my private SSH key into the text box - connected successfully
- Created an application from this repository via UI - success
- Synced the application - success

Can you please exactly point out the steps you are doing? Do you maybe have a C&P error in your SSH private key data?

**Comment by scbizu** [comment]
> > And I think that I catch the bug :
> > As https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L1151 uses the `getCAPath` to fetch the SSH Cred file path , but the `getCAPath` only checks if it is a HTTPS Cred and no options for SSH .
> 
> Hm, that function only resolves to the file the TLS CA certificate for HTTPS connected repositories should be loaded from. The SSH private key is actually stored in a Kubernetes secret (and written out to a temporary file during connection phase).
> 
> I just tried to reproduce the following:
> 
> * Connect new repository via SSH using the UI, using `git@gitlab.com:user/repo` as the repository URL and pasted my private SSH key into the text box - connected successfully
> * Created an application from this repository via UI - success
> * Synced the application - success
> 
> Can you please exactly point out the steps you are doing? Do you maybe have a C&P error in your SSH private key data?

I can successfully connect to the my gitlab repo (the first step) , and failed at the second step when I try to create the app from my connected repo . Maybe I can give you some pod logs to help debugging ?

**Comment by davidkarlsen** [comment]
See the error message - the port is lost:
```
argocd app create --auto-prune --dest-namespace eos-jfr-srv-d1 --directory-recurse --name tac-jfr --path argocd --revision feature/argocd --repo ssh://git@fsstash.evry.com:7999/tac_jfr/tac-jfr-server.git --sync-policy automated 
FATA[0000] rpc error: code = InvalidArgument desc = application spec is invalid: InvalidSpecError: Unable to get app details: rpc error: code = Internal desc = Failed to fetch git repo: `git fetch origin --tags --force` failed exit status 128: Warning: Permanently added 'fsstash.evry.com,10.252.176.59' (ECDSA) to the list of known hosts.%0D%0APermission denied, please try again.%0D%0APermission denied, please try again.%0D%0Agit@fsstash.evry.com: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).%0D%0Afatal: Could not read from remote repository.%0A%0APlease make sure you have the correct access rights%0Aand the repository exists. 
```
no port 7999 there

**Comment by davidkarlsen** [comment]
Indeed if I tcpdump on the host I see packets going to port 22.

**Comment by scbizu** [comment]
@davidkarlsen maybe you are right , my gitlab server is not at the default 22 port.

**Comment by davidkarlsen** [comment]
also see https://github.com/argoproj/argo-cd/issues/2576

**Comment by jannfis** [comment]
Thanks for the additional info, @davidkarlsen - I'll setup a test repository on non-default port for further investigation. Long vacation seem to have erased some memory, I forgot about the issue I have opened myself ;-)

**Comment by jannfis** [comment]
Hm, actually, all the end to end tests related to Git via SSH already run against a non-default SSH repository (2222).

**Comment by davidkarlsen** [comment]
There seems to be several implementations of the ssh-url parsing, like https://github.com/argoproj/argo-cd/pull/2948 for instance.

**Comment by davidkarlsen** [comment]
https://argoproj.slack.com/archives/CASHNF6MS/p1579009775192100?thread_ts=1579003422.187400&cid=CASHNF6MS

**Comment by jannfis** [comment]
> There seems to be several implementations of the ssh-url parsing, like #2948 for instance.

Yep, there's something fishy about it. 

Can you actually create the application using the UI, i.e. by navigating to the repositories page, click the three vertical dots next to the repo and select "Add application"?

**Comment by davidkarlsen** [comment]
nope: 
<img width="1441" alt="Screenshot 2020-01-21 at 19 16 41" src="https://user-images.githubusercontent.com/18299/72831195-a29f8600-3c82-11ea-9d7e-0bc73cfac3a0.png">

**Comment by jannfis** [comment]
Hm, in your screenshot, the URL is `git@fsstash.evry.com:7999/...` while it should really be `ssh://@fsstash.evry.com:7999/...` - scp-style target destinations do not support a port specification, while `ssh://`-type URLs do.

Your repository is defined as `git@fsstash.evry.com:7999/tac_jfr/tac-jfr-server.git` I suppose. While testing the connection, it is internally converted to `ssh://git@fsstash.evry.com:7999/tac_jfr/tac-jfr-server.git` (so the connection test succeeds), but when used for i.e. Git, it is used verbatim - the port specification actually is part of the path, and default port 22 is used.

This is a bug, no question. As a workaround, please remove your configured repository and re-add it using `ssh://` URL-scheme. This should work (and I have just tested it on a local setup that it does).

**Comment by ppmathis** [comment]
I've experienced the same issue with ArgoCD 1.4.2, both when using rsa and ed25519 keys. While adding/connecting the repository worked and showed up as successful, I was not able to add any application. Additionally, the path selector in the new application dialog did not work, probably for the same reason. As an additional remark, I've used the `git@<server>` syntax and Git is running on the default SSH port (tcp/22), but the UI wizard still did not work.

The workaround mentioned by @philippeboyd, which is adding the repository and its SSH private key through CLI, worked for me though.

**Comment by jannfis** [comment]
@ppmathis What is the exact error messages? I just tried the following using UI on 1.4.2 and it works. There's definitely a bug somewhere, but I need to be able to reproduce it. So every piece of concrete information that could be valuable for solving this bug is much appreciated.

![image](https://user-images.githubusercontent.com/3942683/73075348-fd6ff200-3ebb-11ea-9e8e-ece96ae19930.png)

![image](https://user-images.githubusercontent.com/3942683/73075524-68212d80-3ebc-11ea-862b-f611e8a18f98.png)

![image](https://user-images.githubusercontent.com/3942683/73075633-a61e5180-3ebc-11ea-82ee-33d67f325d85.png)

**Comment by ppmathis** [comment]
@jannfis Thank you for your response and sorry, I missed mentioning that I am receiving the same error message which has been originally posted on this issue:

> Unable to create application: application spec is invalid: InvalidSpecError: Unable to get app details: rpc error: code = Internal desc = Failed to fetch git repo: `git fetch origin --tags --force` failed exit status 128: Failed to add the RSA host key for IP address '<repo server>' to the list of known hosts (/app/config/ssh/ssh_known_hosts). Load key "/dev/shm/422544524": invalid format git@<repo server>: Permission denied (publickey). fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.

To help with reproducing this specific issue, I've created two new repositories `argocd-gh2890-1` and `argocd-gh2890-2`, created a 4096 bit RSA key using `ssh-keygen` and added the same public key to both repositories as a deployment key (= read-only repository access). I've then proceeded by pushing the same contents to both repositories and adding them in separate ways:

1) `argocd-gh2890-1` has been added through the UI, ensuring that no spaces are present before or after the private SSH key which I've copy-pasted from the terminal
2) `argocd-gh2890-2` has been added through the CLI using this command: `argocd repo add git@<repo server>:ppmathis/argocd-gh2890-2.git --ssh-private-key-path argocd --name gh2890-2`

So far the behavior does not differ and both repositories show up as successfully connected. When trying to add a new application, the outcome is different:

1) `argocd-gh2890-1` suffers from the issue I mentioned and throws the error message I quoted in the beginning of this issue. The path selector does not show any suggestions and adding the application fails.
2) `argocd-gh2890-2` works just fine. The path selector shows all recognized kustomize deployment files and adding the application works.

I continued by extracting the secrets into separate files and analyzed them using xxd and vimdiff:

```shell
kubectl -n argocd get secrets -o go-template --template '{{.data.sshPrivateKey}}' repo-argocd-gh2890-1-3746518547 | base64 -d > secret1
kubectl -n argocd get secrets -o go-template --template '{{.data.sshPrivateKey}}' repo-argocd-gh2890-2-163780834 | base64 -d > secret2
vimdiff <(xxd secret1) <(xxd secret2)
```

![vimdiff of k8s secrets](https://zig.pw/d53fc234e8a0afe834557ef1a5cda3f7065563d1.png)

The devil is in the detail as the screenshot clearly shows, as the secret added through the CLI (`secret2` - right side) has a newline at the end whereas the key added through the WebUI does not. This also explains why the issue only occurs for some people when using the WebUI - when copying from a terminal (e.g. `cat` + manual selection), the newline usually does not get copied, however when copying from a graphical text editor the newline will be part of the clipboard.

To confirm the issue, I've added a third repository with the same contents named `argocd-gh2890-3` with the same deployment key and added it through the UI once again, ensuring that the private key has a newline at the end - and everything works smoothly.

**TL;DR:** The SSH private key must have a single newline character at the end. While both cases (with and without) work just fine for the repository status check, the actual `git fetch` command fails without a trailing newline.

**Comment by jannfis** [comment]
Thank you very much for this excellent analysis, @ppmathis! 

The different behaviour is most likely due to the fact that the repository connection check is done using the go-git Git native implementation in Git, while most other Git operations (such as fetch) are performed using the Git client binary, which in turn tunnels the connection through OpenSSH. Indeed, the different behaviour between `go-git` and `git` could be the culprit for many of the issues regarding "repository connection works when adding the repo" (performed by go-git) and "same repository connection does not work when adding application" (performed by git binary).

I guess the temporary fix for this issue would be to ensure that the private key has a trailing newline before being stored in the secret, without being dependent whether the API client (CLI, UI, custom, ...) actually sends a trailing newline along, while the long term fix would involve some refactoring of the Git code in ArgoCD and deciding on a single implementation to use.

I will perform a few more tests and send a fix shortly. Again, thank you very much for the analysis, it's much appreciated!

**Comment by jannfis** [comment]
So just for the record: For any version not having the fix from #3064 included, the workaround is to a) either make sure that the SSH private key in the UI has a trailing new-line character or to add the private key via CLI.

**Comment by flypolarbear** [comment]
> > So just for the record: For any version not having the fix from #3064 included, the workaround is to a) either make sure that the SSH private key in the UI has a trailing new-line character or to add the private key via CLI.
> 
> Hi guys, neither of these solutions are working for me
> 
> I still can manage to connect on repo, but, when i add the project, i get the error
> 
> ```
> rpc error: code = Internal desc = Failed to fetch f708f3874e6cb4000018a83d17d0221f0a053dba: `git fetch origin --tags --force` failed exit status 128: No user exists for uid 1001 fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.
> 8 minutes ago (Thu Oct 28 2021 11:42:17 GMT-0300)
> ```
> 
> I'm using argo 2.0.4. Does anybody have another solution?

I encountered the same problem in version 2.1.1. Did you solve it?

**Comment by yngvark** [comment]
I'm encountering this in 2.1.7. Not sure if the problems is if I did something wrong or ArgoCD yet.

**Comment by raphaelLacerda** [comment]
> > > So just for the record: For any version not having the fix from #3064 included, the workaround is to a) either make sure that the SSH private key in the UI has a trailing new-line character or to add the private key via CLI.
> > 
> > 
> > Hi guys, neither of these solutions are working for me
> > I still can manage to connect on repo, but, when i add the project, i get the error
> > ```
> > rpc error: code = Internal desc = Failed to fetch f708f3874e6cb4000018a83d17d0221f0a053dba: `git fetch origin --tags --force` failed exit status 128: No user exists for uid 1001 fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.
> > 8 minutes ago (Thu Oct 28 2021 11:42:17 GMT-0300)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm using argo 2.0.4. Does anybody have another solution?
> 
> I encountered the same problem in version 2.1.1. Did you solve it?

Actually i did!

I'm using docker, helm chart and kubernetes to deploy Argo.

So, i've created a custom argo imagem that contains my ssh keys stores at /root/.ssh


`FROM argo-cd:2.0.4
USER root

COPY id_rsa /root/.ssh/id_rsa
COPY id_rsa.pub /root/.ssh/id_rsa.pub
COPY known_hosts /root/.ssh/known_hosts

RUN chmod 400 /root/.ssh/id_rsa
RUN chmod 400 /root/.ssh/id_rsa.pub
RUN chmod 400 /root/.ssh/known_hosts

`


Dont't forget to run CHMOD 400

So it worked beautifully

**Comment by kreeuwijk** [comment]
I was able to fix this in Kubernetes by setting `runAsUser` to a valid userID that existed in the argo-cd-repo-server container:
```
       securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - all
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          runAsUser: 65534
```

**Comment by Vavinash-github** [comment]
I am connecting to google cloud source repository with the help of workload Identity ---> 
![image](https://github.com/argoproj/argo-cd/assets/144169783/a8ca47de-05cc-4ff4-9f10-89d231d86fd7)

I am successfully able to connect to repository using both UI and CLI but when I create an Application The path selector does not show any suggestions and adding the application fails.

### Argocd Version: 2.8.2
### Error:
finished unary call with code Internal" error="rpc error: code = Internal desc = Failed to fetch default: `git fetch origin --tags --force --prune` failed exit status 128: remote: PERMISSION_DENIED: The caller does not have permission\nremote: [type.googleapis.com/google.rpc.RequestInfo]\nremote: request_id: \"86f202a3b8eba2e8e40bb586\"\nfatal: Authentication failed for 'https://source.developers.google.com/p/projectid/r/reponame/'" grpc.code=Internal grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-09-07T03:42:48Z" grpc.time_ms=2607.152 span.kind=server system=grpc

**Comment by karan6190** [comment]
> I am connecting to google cloud source repository with the help of workload Identity ---> ![image](https://user-images.githubusercontent.com/144169783/266197228-a8ca47de-05cc-4ff4-9f10-89d231d86fd7.png)
> 
> I am successfully able to connect to repository using both UI and CLI but when I create an Application The path selector does not show any suggestions and adding the application fails.
> 
> ### Argocd Version: 2.8.2
> ### Error:
> finished unary call with code Internal" error="rpc error: code = Internal desc = Failed to fetch default: `git fetch origin --tags --force --prune` failed exit status 128: remote: PERMISSION_DENIED: The caller does not have permission\nremote: [type.googleapis.com/google.rpc.RequestInfo]\nremote: request_id: "86f202a3b8eba2e8e40bb586"\nfatal: Authentication failed for 'https://source.developers.google.com/p/projectid/r/reponame/'" grpc.code=Internal grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-09-07T03:42:48Z" grpc.time_ms=2607.152 span.kind=server system=grpc

Hey, were you able to resolve the issue?? I am facing the same problem

**Comment by Vavinash-github** [comment]
@karan6190  yes I was able to resolve the issue using the solutions.. Added a dummy client_email in the private key file.

**Comment by karan6190** [comment]
> @karan6190 yes I was able to resolve the issue using the solutions.. Added a dummy client_email in the private key file. Got help from @torfjor in the issue link. Thanks..

Hey, can you please provide the sample, how you solved it ?
